Stitch remote iframe layers together regardless of creation order

achristensen07 opened the pull request:
#### 368b3354ef93a9e982bb7b4cb598de4f1dbe0104
<pre>
Stitch remote iframe layers together regardless of creation order
<a href="https://bugs.webkit.org/show_bug.cgi?id=252009">https://bugs.webkit.org/show_bug.cgi?id=252009</a>
rdar://105236763

Reviewed by Simon Fraser.

In 259937@main we started adding iframe root layers as sublayers of
the hosting layer from another process, but this only works if the hosting
layer is created before the hosted layer.  This adds symmetric code that
remembers the hosting layers and the hosted layers and adds them if the
parent is created first or if the child is created first.

Not quiiiite ready for automated tests covering this, but we&apos;re getting close.
The test added in 259937@main would cover this, but before re-landing that
test we need to get layers from all subframe processes in WebKitTestRunner
and fix at least 1 more assertion.  This can easily be seen in MiniBrowser
with site isolation after resizing the window.  Cross-origin iframes would
disappear before this and now they don&apos;t.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree):
(WebKit::RemoteLayerTreeHost::layerWillBeRemoved):
(WebKit::RemoteLayerTreeHost::createLayer):

Canonical link: <a href="https://commits.webkit.org/260090@main">https://commits.webkit.org/260090@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb5b1deeaf744287444ce861195438abfae5207c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107000 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16052 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39797 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116178 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17518 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7253 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99181 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112765 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96274 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40858 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27916 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82607 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/9173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/29268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9762 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/15337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48827 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11300 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3770 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
```diff
@@ -96,6 +96,7 @@ class RemoteLayerTreeHost {
     RemoteLayerTreeDrawingAreaProxy* m_drawingArea { nullptr };
     WeakPtr<RemoteLayerTreeNode> m_rootNode;
     HashMap<WebCore::GraphicsLayer::PlatformLayerID, std::unique_ptr<RemoteLayerTreeNode>> m_nodes;
+    HashMap<WebCore::LayerHostingContextIdentifier, WebCore::GraphicsLayer::PlatformLayerID> m_hostingLayers;
```
How long do entries stay in these two hashes? Is there a point where we can assert that they are empty?

achristensen07 replied:
Until the layers are removed.  We can basically only assert that they are empty when we have no layers, which doesn't seem useful.

achristensen07 committed:
Stitch remote iframe layers together regardless of creation order
https://bugs.webkit.org/show_bug.cgi?id=252009
rdar://105236763

Reviewed by Simon Fraser.

In 259937@main we started adding iframe root layers as sublayers of
the hosting layer from another process, but this only works if the hosting
layer is created before the hosted layer.  This adds symmetric code that
remembers the hosting layers and the hosted layers and adds them if the
parent is created first or if the child is created first.

Not quiiiite ready for automated tests covering this, but we're getting close.
The test added in 259937@main would cover this, but before re-landing that
test we need to get layers from all subframe processes in WebKitTestRunner
and fix at least 1 more assertion.  This can easily be seen in MiniBrowser
with site isolation after resizing the window.  Cross-origin iframes would
disappear before this and now they don't.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree):
(WebKit::RemoteLayerTreeHost::layerWillBeRemoved):
(WebKit::RemoteLayerTreeHost::createLayer):

Canonical link: https://commits.webkit.org/260090@main
